File: 766-flatten-a-multilevel-doubly-linked-list/flatten-a-multilevel-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/

class Solution {
    private Node prev = null;

    private void flattenList(Node node){
        if(node == null){
            return;
        }

        flattenList(node.next);
        flattenList(node.child);
        node.next = prev;
        if (prev != null){
            prev.prev = node;
        }
        node.child = null;
        prev = node;
    }

    public Node flatten(Node head) {
        flattenList(head);
        return head;
    }
}

File: 766-flatten-a-multilevel-doubly-linked-list/flatten-a-multilevel-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        self.prev = None
        def traverse(node):
            if not node:
                return 
            traverse(node.next)
            traverse(node.child)

            node.next = self.prev
            if self.prev:
                self.prev.prev = node
            node.child = None
            self.prev = node
        
        traverse(head)
        return head